plugins/bcli.c

```diff
@@ -478,23 +492,89 @@ static struct command_result *process_estimatefee(struct bitcoin_cli *bcli)
 
 	feeratetok = json_get_member(bcli->output, tokens, "feerate");
 	if (feeratetok &&
-	    !json_to_bitcoin_amount(bcli->output, feeratetok, &feerate)) {
+	    !json_to_bitcoin_amount(bcli->output, feeratetok, feerate)) {
 		err = tal_fmt(bcli->cmd, "%s: bad 'feerate' field (%.*s)",
 			      bcli_args(bcli), (int)bcli->output_bytes,
 			      bcli->output);
 		return command_done_err(bcli->cmd, BCLI_ERROR, err, NULL);
-	}
+	} else if (!feeratetok)
+		/* We return null if estimation failed, and bitcoin-cli will
+		 * exit with 0 but no feerate field on failure. */
+		return estimatefees_null_response(bcli);
+
+	return NULL;
+}
+
+/* We got all the feerates, give them to lightningd. */
+static struct command_result *estimatefees_final_step(struct bitcoin_cli *bcli)
+{
+	struct command_result *err;
+	struct json_stream *response;
+	struct estimatefees_stash *stash = bcli->stash;
+
+	/* bitcoind could theoretically fail to estimate for a higher target. */
+	if (*bcli->exitstatus != 0)
+		return estimatefees_null_response(bcli);
+
+	err = estimatefees_parse_feerate(bcli, &stash->slow);
+	if (err)
+		return err;
 
-end:
 	response = jsonrpc_stream_success(bcli->cmd);
-	if (feeratetok)
-		json_add_u64(response, "feerate", feerate);
-	else
-		json_add_null(response, "feerate");
+	json_add_u64(response, "urgent", stash->urgent);
+	json_add_u64(response, "normal", stash->normal);
+	json_add_u64(response, "slow", stash->slow);
 
 	return command_finished(bcli->cmd, response);
 }
 
+/* We got the response for the normal feerate, now treat the slow one. */
+static struct command_result *estimatefees_third_step(struct bitcoin_cli *bcli)
+{
+	struct command_result *err;
+	struct estimatefees_stash *stash = bcli->stash;
+	const char **params = tal_arr(bcli->cmd, const char *, 2);
+
+	/* bitcoind could theoretically fail to estimate for a higher target. */
+	if (*bcli->exitstatus != 0)
+		return estimatefees_null_response(bcli);
```

Comment:
why not return what you've gotten so far and log the anomaly?

Reply:
Hmm I could log in `estimatefees_null_response()`, but the fee estimation failing is really common, moreover this one really is only a sanity check as `bitcoin-cli` would return `0` on fee estimation failure..